dbms/src/DataStreams/FilterTransformAction.cpp

```diff
@@ -0,0 +1,181 @@
+        *  and now - are calculated. That is, not all cases are covered by the code above.
+        * This happens if the function returns a constant for a non-constant argument.
+        * For example, `ignore` function.
+        */
+    constant_filter_description = ConstantFilterDescription(*column_of_filter);
+
+    if (constant_filter_description.always_false)
+    {
+        block.clear();
+        return true;
+    }
+
+    IColumn::Filter * filter;
+    ColumnPtr filter_holder;
+
+    if (constant_filter_description.always_true)
+    {
+        return true;
+    }
+    else
+    {
+        FilterDescription filter_and_holder(*column_of_filter);
+        filter = const_cast<IColumn::Filter *>(filter_and_holder.data);
+        filter_holder = filter_and_holder.data_holder;
+    }
+
+    /** Let's find out how many rows will be in result.
+      * To do this, we filter out the first non-constant column
+      *  or calculate number of set bytes in the filter.
+      */
+    size_t first_non_constant_column = 0;
+    for (size_t i = 0; i < columns; ++i)
+    {
+        if (!block.safeGetByPosition(i).column->isColumnConst())
+        {
+            first_non_constant_column = i;
+
+            if (first_non_constant_column != filter_column)
+                break;
+        }
+    }
+
+    size_t filtered_rows = 0;
+    if (first_non_constant_column != filter_column)
+    {
+        ColumnWithTypeAndName & current_column = block.safeGetByPosition(first_non_constant_column);
+        current_column.column = current_column.column->filter(*filter, -1);
+        filtered_rows = current_column.column->size();
+    }
+    else
+    {
+        filtered_rows = countBytesInFilter(*filter);
+    }
+
+    /// If the current block is completely filtered out, let's move on to the next one.
+    if (filtered_rows == 0)
+        return false;
+
+    /// If all the rows pass through the filter.
+    if (filtered_rows == rows)
+    {
+        /// Replace the column with the filter by a constant.
+        block.safeGetByPosition(filter_column).column
+            = block.safeGetByPosition(filter_column).type->createColumnConst(filtered_rows, UInt64(1));
+        /// No need to touch the rest of the columns.
+        return true;
+    }
+
+    /// Filter the rest of the columns.
+    for (size_t i = 0; i < columns; ++i)
+    {
+        ColumnWithTypeAndName & current_column = block.safeGetByPosition(i);
+
+        if (i == filter_column)
+        {
+            /// The column with filter itself is replaced with a column with a constant `1`, since after filtering, nothing else will remain.
+            /// NOTE User could pass column with something different than 0 and 1 for filter.
+            /// Example:
+            ///  SELECT materialize(100) AS x WHERE x
+            /// will work incorrectly.
+            current_column.column = current_column.type->createColumnConst(filtered_rows, UInt64(1));
+            continue;
+        }
```

Comment:
Why always replace the filter column with a constant `1`? What if `select a from t where a > 0`?

Reply:
Because after filter, the value of filter_column must be `1`

Reply:
if the data type of `a` isn't bool, the filter column will be `tidb_cast(a, bool)`.

Reply:
so, for `select a from t where a > 0`, there are two columns in a block, `a` and `tidb_cast(a, bool)`?

Reply:
yes, the related code is https://github.com/pingcap/tiflash/blob/9d7d060e67637e5c1f8854f6ba5e7b89d8570eae/dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp#L680-L709

Reply:
get it, thanks